include/pmacc/particles/operations/splitIntoListOfFrames.kernel

```diff
@@ -47,192 +56,402 @@ namespace kernel
+            using namespace mappings::threads;
 
-            typedef T_SrcFrame SrcFrameType;
-            typedef typename T_DestBox::FrameType DestFrameType;
-            typedef typename T_DestBox::FramePtr DestFramePtr;
-            typedef typename T_CellDescription::SuperCellSize SuperCellSize;
-            constexpr unsigned NumDims = T_DestBox::Dim;
-            constexpr uint32_t particlesPerFrame = pmacc::math::CT::volume<SuperCellSize>::type::value;
+            using SrcFrameType = T_SrcFrame;
+            using DestFrameType = typename T_DestBox::FrameType;
+            using DestFramePtr = typename T_DestBox::FramePtr;
+            using SuperCellSize = typename DestFrameType::SuperCellSize;
 
-            PMACC_SMEM( destFramePtr, memory::Array< DestFramePtr, particlesPerFrame > );
-            PMACC_SMEM( linearSuperCellIds, memory::Array< int, particlesPerFrame > );
-            PMACC_SMEM( srcParticleOffset, int );
+            constexpr uint32_t numWorkers = T_numWorkers;
+            constexpr uint32_t numDims = T_DestBox::Dim;
+            constexpr uint32_t particlesPerFrame = math::CT::volume< SuperCellSize >::type::value;
 
+            PMACC_SMEM(
+                destFramePtr,
+                memory::Array<
+                    DestFramePtr,
+                    particlesPerFrame
+                >
+            );
+            PMACC_SMEM(
+                sharedLinearSuperCellIds,
+                memory::Array<
+                    int,
+                    particlesPerFrame
+                >
+            );
+            PMACC_SMEM(
+                sharedSrcParticleOffset,
+                int
+            );
 
-            const int linearThreadIdx = threadIdx.x;
+            uint32_t const workerIdx = threadIdx.x;
 
+            DataSpace< numDims > const numSuperCells(
+                cellDesc.getGridSuperCells( ) - cellDesc.getGuardingSuperCells( ) * 2
+            );
 
-            const DataSpace<NumDims> numSuperCells(cellDesc.getGridSuperCells() - cellDesc.getGuardingSuperCells()*2);
-            if (linearThreadIdx == 0)
-            {
-                /* apply for work for the full block
-                 * counter [0] -> offset to load particles
-                 */
-                srcParticleOffset = atomicAdd(&(counter[0]), particlesPerFrame);
-            }
-            destFramePtr[linearThreadIdx] = DestFramePtr();
-            linearSuperCellIds[linearThreadIdx] = -1;
+            ForEachIdx<
+                IdxConfig<
+                    1,
+                    numWorkers
+                >
+            > onlyMaster{ workerIdx };
 
-            __syncthreads();
+            onlyMaster(
+                [&](
+                    uint32_t const,
+                    uint32_t const
+                )
+                {
+                    /* apply for work for the full block
+                     * counter [0] -> offset to load particles
+                     */
```

Comment:
At least the comment could remind that `counter[0]` is `sharedSrcParticleOffset`. Furthermore the comment is not very helpful. I am adding the particles per frame to an offset... What?